man/man1/x2goclient.1
'\" -*- coding: utf-8 -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH x2goclient 1 "Jun 2023" "Version 4.1.2.4" "X2Go Client (Qt4)"
.SH NAME
x2goclient \- Client application to launch server-side X2Go sessions.
.SH SYNOPSIS
'nh
.fi
.ad l
\fBx2goclient\fR  <options>

.SH DESCRIPTION
\fBx2goclient\fR is a GUI application for launching server-side X2Go sessions.
.PP
.SH HELP AND DEBUG OPTIONS
\fBx2goclient\fR has the following help and debugging options:
.TP
\*(T<\fB\-\-help\fR\*(T>
Show available command line options.
.TP
\*(T<\fB\-\-help\-pack\fR\*(T>
Show available pack methods and exit.
.TP
\*(T<\fB\-v, \-\-version\fR\*(T>
Output X2Go Client version to stderr.
.TP
\*(T<\fB\-\-debug\fR\*(T>
Enable extensive console output for debugging X2Go Client.
.TP
\*(T<\fB\-\-changelog\fR\*(T>
Show changelog file.
.TP
\*(T<\fB\-\-git\-info\fR\*(T>
Show Git commit history (not available on all X2Go Client builds).
.SH GUI OPTIONS
The following options can be used for controlling the GUI behaviour:
.TP
\*(T<\fB\-\-no\-menu\fR\*(T>
Hide menu-/toolbar (default: false).
.TP
\*(T<\fB\-\-maximize\fR\*(T>
Start client maximized (default: false).
.TP
\*(T<\fB\-\-hide\fR\*(T>
Hide client (start hidden, default: false).
.TP
\*(T<\fB\-\-keep\-trayicon\fR\*(T>
Always keep the "accelerated X" icon in the system tray. Without this option the default behaviour is:
Switch the tray icon to the session's icon on successful session startup/resumption.
.TP
\*(T<\fB\-\-tray\-icon\fR\*(T>
Show tray icon (default: false). This option overrides the global tray icon settings and also forces the tray icon
to be shown in hidden X2Go Client mode.
.TP
\*(T<\fB\-\-hide\-foldersharing\fR\*(T>
Hide all elements in X2Go Client GUI that are related to folder sharing.
.TP
\*(T<\fB\-\-close\-disconnect\fR\*(T>
Close X2Go Client after a session got disconnected.
.TP
\*(T<\fB\-\-client\-ssh\-port\fR\*(T>
Local ssh port (for filesystem export, default: 22).
.TP
\*(T<\fB\-\-autostart=<app>\fR\*(T>
Launch <app> by session start in "published applications" mode. This parameter should go along
with --session=<session-profile-name>, where the given session profile should be configured as
a session profile of type "Published Applications".
.TP
\*(T<\fB\-\-support\-menu=<html\-file>\fR\*(T>
Add a "Support..." submenu item to X2Go Client's "Help" menu.

The support information is provided via a simple HTML file. Example:
.sp
.if n \{\
.RS 4
.\}
.nf
        \fI<html><body>
        <h1>IT-Support for X2Go</h1>
        DAS-NETZWERKTEAM | Mike Gabriel:<br />
        <a href="mailto:support@das-netzwerkteam.de">
            support@das-netzwerkteam.de
        </a><br />
        <br />
        Thanks for using X2Go...
        </body></html>\fR
.fi
.if n \{\
.RE
.\}
.TP
\*(T<\fB\-\-background=<svg\-file|dir>\fR\*(T>
Use a custom/branded background image (SVG format) for X2Go Client's main window (i.e. replace the blue gradient area
in X2Go Client by some background of your own (business/workgroup/project).
If a directory is given, will randomly pick an SVG file inside of it.
.TP
\*(T<\fB\-\-branding=<svg\-file>\fR\*(T>
Use a custom icon (SVG format) for additional branding to replace the default in the lower left corner of X2Go Client's main window.

.SH MISCELLANEOUS OPTIONS
These options change X2Go Client's behavior globally. Some might not be available on all platforms.
.TP
\*(T<\fB\-\-xserver-start-limit=<num>\fR\*(T> (Windows-only)
Limit the X.Org Server start count to at most <num> tries. Default: \fB3\fR.
If \fB<num>\fR is zero or a negative number, no limit is imposed.
The limits for \fB<num>\fR are platform dependent. If the passed value is out of bounds, X2Go Client falls back to the default value.

.SH SESSION PROFILE DEFAULTS
You can set certain defaults for session profiles via command line switches using the following options:
.TP
\*(T<\fB\-\-user\fR\*(T>
Pre-selection of user at client startup (LDAP mode only).
.TP
\*(T<\fB\-\-command=<default\-command>\fR\*(T>
Default command for session startup.
.TP
\*(T<\fB\-\-ssh\-port\fR\*(T>
Use this default TCP/IP port for connection (default: 22).
.TP
\*(T<\fB\-\-ssh\-key\fR\*(T>
Use this SSH private key file by default.
.TP
\*(T<\fB\-\-autologin\fR\*(T>
By default, search for standard SSH key file names or try using an SSH agent.
.TP
\*(T<\fB\-\-link\fR\*(T>
Set default link type (modem,isdn,adsl,wan or lan, default: adsl).
.TP
\*(T<\fB\-\-pack\fR\*(T>
Set default pack method (default: '16m-jpeg').
.TP
\*(T<\fB\-\-quality\fR\*(T>
Set default image quality(0-9, default: 9).
.TP
\*(T<\fB\-\-set\-kbd\fR\*(T>
Overwrite current keyboard settings, no override by default.
.TP
\*(T<\fB\-\-kbd\-layout\fR\*(T>
Set keyboard layout (default: 'de').
.TP
\*(T<\fB\-\-kbd\-type\fR\*(T>
Set keyboard type (default: pc105/us, depend on our clients language setup, though).
.TP
\*(T<\fB\-\-clipboard=<both|client|server|none>\fR\*(T>
Set default clipboard mode.
.TP
\*(T<\fB\-\-geometry=<W>x<H>|fullscreen\fR\*(T>
Set default geometry, default '800x600'.
.TP
\*(T<\fB\-\-dpi=<dpi>\fR\*(T>
set DPI of X2Go Agent to <dpi>, default set to same as local display.
.TP
\*(T<\fB\-\-sound\fR\*(T>
Activate sound for session, not enabled by default.
.TP
\*(T<\fB\-\-sound\-system\fR\*(T>
Which soundsystem to use: arts, esd, pulse (default: pulse).
.TP
\*(T<\fB\-\-xinerama\fR\*(T>
Use Xinerama by default.
.TP
\*(T<\fB\-\-session\-icon=<128x128px\-icon\-png\-file>\fR\*(T>
Use a custom session icon by default.

.SH THIN CLIENT OPTIONS
The following command line options are primarily interesting if \fBx2goclient\fR is used
as a login manager on X2Go thin clients.
.TP
\*(T<\fB\-\-thinclient\fR\*(T>
Let X2Go Client run without window manager.
.TP
\*(T<\fB\-\-sessionid=<session\-profile-id>\fR\*(T>
Pre-selection of a session profile (by its session profile ID) at client startup.
.TP
\*(T<\fB\-\-session=<session\-profile\-name>\fR\*(T>
Pre-selection of a session profile (by its session profile name) at client startup.
.TP
\*(T<\fB\-\-no\-session\-edit\fR\*(T>
Disable session editing.
.TP
\*(T<\fB\-\-pgp\-card\fR\*(T>
Use openPGP Card authentication (default: false).
.TP
\*(T<\fB\-\-external\-login=<smartcard\-logins>\fR\*(T>
Authenticate via SMART card, path to login notification file.
.TP
\*(T<\fB\-\-add\-to\-known\-hosts\fR\*(T>
Add DSA/RSA host key fingerprint to .ssh/known_hosts in case of "authenticity of server can't be established".
.TP
\*(T<\fB\-\-read\-exports\-from\fR\*(T>
Specifies a directory where some external mechanism (e.g. script) can notify \fBx2goclient\fR on
new block devices (CD/DVDs, USB sticks, etc.).
.TP
\*(T<\fB\-\-haltbt\fR\*(T>
Show a "Shutdown" button in the top left corner of X2Go Client. When pressed, X2Go Client will create the file
$HOME/.halt. An external script has to watch for this file and trigger a thinclient shutdown once it appears.

.SH PORTABLE OPTIONS
X2Go Client can be installed on a USB-Flashdrive as a portable applications. If installed that way, the following
options support the portable setup of X2Go Client:
.TP
\*(T<\fB\-\-portable\fR\*(T>
Start X2Go Client in "portable" mode. This implies... (TODO).
.TP
\*(T<\fB\-\-home=<dir>\fR\*(T>
Set the users home directory. This is the directory where X2Go Client will store settings and cache files.
.TP
\*(T<\fB\-\-session\-conf=<file>\fR\*(T>
Path to alternative session profile config (default: $HOME/.x2goclient/sessions).
.TP
\*(T<\fB\-\-clean\-all\-files\fR\*(T>
In portable mode, X2Go Client removes several directories it has created (and probably directories you
don't want to have removed on a normal system):

        \fI$HOME/.ssh
        $HOME/ssh
        $HOME/.x2go\fR

If \fI\-\-clean\-all\-files\fR is used on the command line, also the directory containing the configuration files of
X2Go Client gets removed:

        \fI$HOME/.x2goclient\fR

.SH BROKER OPTIONS
In case you want to retrieve \fBx2goclient\fR session profiles from an X2Go Session Broker use the following options:
.TP
\*(T<\fB\-\-broker\-url=<URL>\fR\*(T>
Specify the <URL> of the X2Go Session Broker. X2Go Client can access http:// and ssh:// style URLs.

Syntax of <URL> for HTTP brokerage:

http(s)://<user>:<password>@<hostname>:<port>/path/to/broker

Syntax of <URL> for  SSH brokerage:

ssh://<user>:<password>@<hostname>:<port>/usr/bin/x2gobroker (or any executable that
provides the broker via SSH).

.TP
\*(T<\fB\-\-broker\-cacertfile=</path/to/cafile.crt>\fR\*(T>
Specify a special (self-signed) root-CACert file that shall get used when connecting to an X2Go Session Broker via https (SSL).
.TP
\*(T<\fB\-\-broker\-noauth\fR\*(T>
The X2Go Session Broker is accessible without authentication.
.TP
\*(T<\fB\-\-broker\-krblogin\fR\*(T>
Login to X2Go Session Broker via GSSAPI. Only supported with ssh:// broker URLs.
.TP
\*(T<\fB\-\-auth\-id=<USERNAME>\fR\*(T>
Use this <USERNAME> for authenticating against the X2Go Session Broker. This option mostly makes sense together
with \fI--broker-autologin\fR or \fI--broker-ssh-key\fR.
.TP
\*(T<\fB\-\-broker\-autologin\fR\*(T>
For SSH based X2Go Session Brokers. If an SSH agent is available or default key files exist then
try those for authentication against the X2Go Session Broker.
.TP
\*(T<\fB\-\-broker\-autologoff\fR\*(T>
Enforce re-authentication against X2Go Session Broker after a session has been suspended or terminated.
.TP
\*(T<\fB\-\-broker\-ssh\-key=<SSHPRIVKEY>\fR\*(T>
For SSH based X2Go Session Brokers. Full path to a valid SSH private key file.
.TP
\*(T<\fB\-\-broker\-name=<NAME>\fR\*(T>
Sets the broker name to display in X2Go Client. This parameter is optional.
.TP
\*(T<\fB\-\-connectivity\-test\fR\*(T>
Run a connectivity test against X2Go Session Broker. Not supported by all broker implementations.
.TP
\*(T<\fB\-\-change\-broker\-pass\fR\*(T>
Send a change-password request to an X2Go Session Broker. Not supported by all broker implementations.
.TP
\*(T<\fB\-\-broker\-noauth\-with\-session\-username\fR\*(T>
Informs the broker to use the username you entered for X2Go Server authentication when selecting a server/session. Only has functionality if --broker-noauth is used.
.TP
\*(T<\fB\-\-broker\-use\-creds\-for\-session\fR\*(T>
Use broker credentials as session user/password credentials when using broker mode with broker authentication. Currently only affects direct RDP sessions.
.TP
\*(T<\fB\-\-broker\-logoutbt\fR\*(T>
Show a "Logout" button in the top left corner of X2Go Client. When logged in to a broker, pressing this button
X2Go Client lets you login to the broker again.

.SH LDAP OPTIONS (deprecated)
NOTE: LDAP support won't be continued in X2Go Client 2 (next generation of X2Go Client).
.PP
A slightly different approach than X2Go Client's LDAP functionality is available via the X2Go Session Brokerage feature.
.PP
In case you want to control \fBx2goclient\fR via an LDAP server use the following options:
.TP
\*(T<\fB\-\-ldap=<host:port:dn>\fR\*(T>
Start with LDAP support (disabled by default). Example: --ldap=ldapserver:389:o=organization,c=de
.TP
\*(T<\fB\-\-ldap1=<host:port:dn>\fR\*(T>
Failover LDAP URI No. 1.
.TP
\*(T<\fB\-\-ldap2=<host:port:dn>\fR\*(T>
Failover LDAP URI No. 2.
.TP
\*(T<\fB\-\-ldap\-printing\fR\*(T>
Allow client side printing in LDAP mode.

.PP
.SH AUTHOR
This manual has been written by Mike Gabriel <mike.gabriel@das-netzwerkteam.de> for the X2Go project
(https://www.x2go.org).


debian/changelog
x2goclient (4.1.2.4-0x2go1) UNRELEASED; urgency=medium

  * Continue development

 -- X2Go Release Manager <git-admin@x2go.org>  Thu, 29 Jun 2023 20:54:03 +0200

x2goclient (4.1.2.3-0x2go1) unstable; urgency=medium

  [ Oleksandr Shneyder ]
  * New upstream version (4.1.2.3):
    - Load alternative image resources from specified folder.
    - Set correct aspect ratio for x2go and ON logos.
    - Download alternative resource files from http server.
    - Support for PNG background images.
    - Display state of broker connection in status bar.
    - Client can synchronize sessions with broker. Broker need to send
      syncinterval value in the client config section.
    - Set min width of session status label.
    - set clipboard mode of x2gokdriveclient in cmd line option.
    - Fix maximizing client window in thinclient mode with Qt5 (bullseye).
    - Add mime type application/x-x2go. X2Go Client can now automatically open
      session files with .x2go extension.
    - X2Go Client will send it's version to the broker when sending client
      event.
    - Don't close X2Go Client if settings dialog is called before broker auth.
    - X2Go Client will send it's OS name to the broker when sending client
      event.
    - Don't exit if connection to HTTP broker is failed when sync sessions.
    - Support for HTTP(S) urls in the session icons when using http broker.
    - Fix loading session icon to system tray when using broker.
    - Add --debug argument to x2gokdriveclient in debug mode.
    - Setting debug filter rules for QT>=5.2
    - Implement different view modes for sessions explorer:
      Favorite sessions, running sessions, all sessions.
    - Improve view modes in broker mode.
    - Set toolTip with detailed information for session button.
    - Show session path in session button in Favorite and Running views.
    - Improve synchronisation in broker mode.
    - Fix loading HTTP icon in pass dialog in broker mode.
    - Replace QFont::Thin with 0 to keep compatibility with Qt4.
    - Add special ACL for exported directories in Windows client.
    - Support for rootless X2GoKdrive sessions.
    - Fixed build with Qt5 on Windows.
    - Do not show empty folders in session explorer.
    - Support for OPENSSH PRIVATE KEY format sent from broker.
    - Do not show client in broker mode if startet in "hidden" mode.
    - Enable/Disable session view buttons inside of sessions explorer.
    - Add splash screen option when starting in hidden mode.
    - replace <HOME> with user's home dir in exports directories,
      when exporting session default directories.
    - create default export directory if not exists.
    - send to broker xdmcpserver in command option when starting xdmcp session.
    - improve processing stderr from nxproxy.
    - support for recent cygwin binaries and nxproxy 3.5.99.x
      all cygwin binaries and DLLs should be moved into the
      INSTDIR/bin directory by Windows installer.
    - for x2gokdrive sessions, when auto kbd layout is chosen,
      x2goclient will try to set the same kbd layout on server as on the
      client.
    - add "noresize" setting, which makes the proxy window not resizable
      for user.
    - on Windows SSHD will only listen 127.0.0.1 to avoid FW warnings.
    - on Windows check the location of cgwin binaries in <APPDIR> and
      in <APPDIR>/bin.

  [ Ryan Schmidt ]
  * New upstream version (4.1.2.3):
    - x2goclient.pro: use the right C++ standard library on OS X/macOS.

  [ Melroy van den Berg ]
  * New upstream version (4.1.2.3):
    - src/sshmasterconnection.c: send EOF before closing channel.
    - src/sshmasterconnection.c: return false if login check remote command
      execution failed.
    - src/sshmasterconnection.c: refactor some of the channel loop and fix
      channel/session closes.
    - src/sshmasterconnection.c: fix 'when not' password auth. Avoids double
      error messages.
    - src/sshmasterconnection.c: correct retval check.
    - src/sshmasterconnection.c: explicitly mark function parameters as unused.
    - .gitignore: ignore Visual Studio temporary files. Fixes: #1469.

  [ Mihai Moldovan ]
  * New upstream version (4.1.2.3):
    - src/sshmasterconnection.{c,h}: fix builds, new function should have been
      registered and not static, but private.
    - .gitignore: add debian/control, since it's now auto-generated.
    - Makefile: support version switching through a QT_VERSION variable -
      might be overridden via environment variables or the make call.
    - x2goclient.spec: add version switching support. Use Qt 5 for FC31+, RHEL
      8 and *SuSE 15.3+.
    - x2goclient.spec: don't try to prepend %{_qt4_bindir} if the macro is not
      defined.
    - Makefile: allow passing down variables that are safe to pass via the
      environment.
    - x2goclient.spec: fix condition with maybe-undefined macro.
    - x2goclient.spec: typo fix in vendor macro.
    - x2goclient.spec: expand %{_qt4_bindir} to integer if it's set, instead
      of using it as a string in the condition (which then leads to parsing
      error).
    - x2goclient.spec: rework Qt5 dependencies. Use a common set of
      PkgConfig()-based dependencies for the Qt 5 libraries themselves and
      only make the linguist tools (lrelease) conditional.
    - Makefile: make controllable variables default-if-not-set to allow
      overriding via the environment.
    - Makefile: invoke make as $(MAKE) to keep original invocation.
    - x2goclient.spec: fix missing %endif in Qt-4-BuildRequires section.
    - x2goclient.spec: use %{_qt5_bindir} macro analogous to the Qt 4 variant.
    - x2goclient.spec: correctly quote PATH overrides.
    - x2goclient.spec: fix syntax error in Qt 5 PATH override.
    - x2goclient.spec: add BuildRequires on Qt5X11Extras.
    - x2goclient.spec: also use/handle %{_libqt5_bindir}, as used by *SuSE.
    - x2goclient.spec: add BuildRequires on Qt5Network. OpenSuSE Tumbleweed

src/x2goclient.cpp
/**************************************************************************
*   Copyright (C) 2005-2023 by Oleksandr Shneyder                         *
*                              <o.shneyder@phoca-gmbh.de>                 *
*   Copyright (C) 2015-2023 by Mihai Moldovan <ionic@ionic.de>            *
*                                                                         *
*   This program is free software; you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation; either version 2 of the License, or     *
*   (at your option) any later version.                                   *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <https://www.gnu.org/licenses/>. *
***************************************************************************/

#include <iostream>
#include <sys/types.h>
#include <unistd.h>
#include <cstring>
#include <cerrno>
#include <cstdlib>
#include <string>
#include <algorithm>
#include <cctype>
#include <vector>
#include <csignal>

#include "unixhelper.h"
#include "ongetpass.h"
#include "compat.h"

int wrap_x2go_main (int argc, char **argv) {
  return (x2goMain (argc, argv));
}

#ifdef Q_OS_UNIX
int fork_helper (int argc, char **argv) {
  /* Fork off to start helper process. */
  pid_t tmp_pid = fork ();

  /* Child. */
  if (0 == tmp_pid) {
    /* Starting unixhelper. */
    std::vector<std::string> new_argv;
    new_argv.push_back (std::string (argv[0]));
    new_argv.push_back ("--unixhelper");

    std::vector<char *> new_argv_c_str;
    for (std::vector<std::string>::iterator it = new_argv.begin (); it != new_argv.end (); ++it) {
      const char *elem = (*it).c_str ();
      new_argv_c_str.push_back (strndup (elem, std::strlen (elem)));
    }

    /* Add null pointer as last element. */
    new_argv_c_str.push_back (0);

    if (0 != execvp (new_argv_c_str.front (), &(new_argv_c_str.front ()))) {
      const int saved_errno = errno;
      std::cerr << "Failed to re-execute process as UNIX cleanup helper tool: " << std::strerror (saved_errno) << "\n"
                << "Terminating and killing parent." << "\n"
                << "Please report a bug, refer to this documentation: https://wiki.x2go.org/doku.php/wiki:bugs" << std::endl;

      pid_t parent_pid = getppid ();
      if (0 != kill (parent_pid, SIGTERM)) {
        const int saved_errno = errno;
        std::cerr << "Failed to kill parent process: " << std::strerror (saved_errno) << std::endl;
      }

      std::exit (EXIT_FAILURE);
    }

    /* Anything here shall be unreachable. */
    return (0);
  }
  /* Error. */
  else if (-1 == tmp_pid) {
    const int saved_errno = errno;
    std::cerr << "Unable to create a new process for the UNIX cleanup watchdog: " << std::strerror (saved_errno) << "\n";
    std::cerr << "Terminating. Please report a bug, refer to this documentation: https://wiki.x2go.org/doku.php/wiki:bugs" << std::endl;

    std::exit (EXIT_FAILURE);
  }
  /* Parent. */
  else {
    /* Start real X2Go Client. */
    return (wrap_x2go_main (argc, argv));
  }
}
#endif /* defined (Q_OS_UNIX) */

int main (int argc, char **argv) {
#ifdef Q_OS_UNIX
  /*
   * Flags we don't need a cleanup helper for, since we know that X2Go Client
   * will never spawn other processes.
   *
   * FIXME: What we'd actually want to have at this point (instead of a
   *        hardcoded list of parameters, anyway) is to use the argument parser
   *        from ONMainWindow (parseParameter). If this function returns false
   *        for any parameter, we know that we won't ever need the UNIX cleanup
   *        helper tool. Sadly, ONMainWindow is only started/available later,
   *        so we can't use any of its functionality here. We'd also need to
   *        make the function side-effect free. It should probably be
   *        refactored into a special options parser class.
   */
  const std::string bypass_flags[] = {
    "--bypass-cleanup-helper",
    "--help",
    "--help-pack",
    "--version",
    "-v",
    "--changelog",
    "--git-info"
  };

  bool bypass_unix_helper = 0;
  bool unix_helper_request = 0;
  for (int i = 0; i < argc; ++i) {
    /* No need to continue scanning if we got the information we were looking for. */
    if ((bypass_unix_helper) && (unix_helper_request)) {
      break;
    }

    std::string cur_arg (argv[i]);

    /* Make the current argument lowercase. */
    std::transform (cur_arg.begin (), cur_arg.end (), cur_arg.begin (), ::tolower);

    if (!(cur_arg.empty ())) {
      /* Scan program arguments for --unixhelper flag. */
      if ((!(unix_helper_request)) && (0 == cur_arg.compare ("--unixhelper"))) {
        unix_helper_request = 1;
      }

      /* Scan for flags bypassing the unix helper. */
      if (!(bypass_unix_helper)) {
        for (std::size_t y = 0; y < (sizeof (bypass_flags) / sizeof (*bypass_flags)); ++y) {
          if (0 == cur_arg.compare (bypass_flags[y])) {
            bypass_unix_helper = 1;
          }
        }
      }
    }
  }

  /* Sanity checks! */
  if ((unix_helper_request) && (bypass_unix_helper)) {
    std::cerr << "Re-execution in UNIX cleanup helper mode was requested, but a command line parameter that is supposed to "
              << "disable the UNIX cleanup helper was found.\n"
              << "Terminating. Please report a bug, refer to this documentation: https://wiki.x2go.org/doku.php/wiki:bugs" << std::endl;

    return (EXIT_FAILURE);
  }

  if (bypass_unix_helper) {
    return (wrap_x2go_main (argc, argv));
  }
  else {
    if (unix_helper_request) {
      /* We were instructed to start as the UNIX cleanup helper tool. */
      return (unixhelper::unix_cleanup (getppid ()));
    }
    else {
      /*
       * setsid() may succeed and we become a session and process
       * group leader, or it may fail indicating that we already
       * are a process group leader. Either way is fine.
       */
      setsid ();

      /* We should be process group leader by now. */
      return (fork_helper (argc, argv));
    }
  }
#else /* defined (Q_OS_UNIX) */
  return (wrap_x2go_main (argc, argv));
#endif /* defined (Q_OS_UNIX) */
}
